### DISCUSSION TITLE
Rust

### DISCUSSION THREAD

**Discussion by Barugon** [godotengine/godot-proposals#4019] [Scripting]
I would really like better Rust integration. I know that there's a new native interface coming for Godot but I feel that Rust should be a first-class language for Godot.

**Comment by kisg** [comment]
You might want to follow these proposals about transforming scripting languages into full blown GDExtensions: #3369 , #3927

While they are not about Rust explicitly, we do touch upon extending the Godot editor to support the Language Server Protocol and the Debug Adapter Protocol to provide better experience for external languages. For compiled languages like C++ or Rust (or C#) we would also need a "Build System Adapter" interface that provides hooks for the different language implementations (e.g. file opened / file saved / run button pressed) so the code can be compiled when necessary. 

Apart from the above Godot Editor extensions, the only thing you need to have for 1st class support of Rust (or any other language really) is a generator that takes the api.json and spits out the necessary interfaces / stubs in Rust.  This is the part where the Godot-Rust project has already made a lot of progress, you should track this issue about Godot 4 and GDExtension support: https://github.com/godot-rust/godot-rust/issues/824

Also tagging @Bromeon could be a good idea... :)

Do you have any specific needs that you would like to have outside of what I described here?

**Comment by azur-wolve** [comment]
By _"first-class language"_ it looks like he means _"official support-integration within the Godot editor"_ aka _"be able to code in rust directly on the scripting workspace"_.

**Reply by Seubmarine** [reply]
You could make an add-on or module to support that pretty easily

**Reply by Xrayez** [reply]
I wouldn't say it's easy, but possible: https://github.com/Vivraan/godot-lang-support

**Comment by Barugon** [comment]
I should probably just wait to see how GDExtension pans out. Currently, using [godot-rust](https://github.com/godot-rust/godot-rust) is painful and clunky. If GDExtension (vs GDNative) can iron out the pain and clunkyness then I suppose there would be no issue.

**Reply by Barugon** [reply]
Still, I would like it to be more directly supported by the Godot team (like .NET is) rather than being a third-rate add-on.

**Reply by Calinou** [reply]
There are no plans to support more programming languages *officially*. GDScript, C#, visual scripting and C++ are a lot of work to support already :slightly_smiling_face:

**Reply by azur-wolve** [reply]
@Barugon 
[this thread](https://github.com/godotengine/godot-proposals/discussions/4007) discusses about both language support and GDExtension
